File: python-backend/models/vasicek.py
"""
Vasicek / CIR — Mean-Reverting Interest Rate Model
Uses RBI Repo Rate instead of US Fed Funds Rate.
Forecasts 12-month INR rate curve for BSM option pricing.
"""

import numpy as np
import logging

logger = logging.getLogger(__name__)


def vasicek_forecast(
    rate_history: list[float],
    horizon_months: int = 12,
    model_type: str = "vasicek",
) -> dict:
    """
    Mean-reverting interest rate forecast on RBI repo rate.

    Args:
        rate_history: List of historical RBI repo rates (% form, e.g. [6.5, 6.5, 6.25, ...])
        horizon_months: Forecast horizon
        model_type: "vasicek" or "cir"

    Returns:
        dict with forecast_curve, long_run_rate, kappa, sigma
    """
    if len(rate_history) < 5:
        return _empty_result("Need at least 5 rate observations")

    try:
        rates = np.array(rate_history, dtype=np.float64) / 100  # to decimal

        # Calibrate parameters using OLS on Δr = a(b - r)Δt + σ ε √Δt
        dt = 1 / 12  # monthly steps
        dr = np.diff(rates)
        r_prev = rates[:-1]

        # OLS: dr = alpha + beta * r_prev + noise
        X = np.column_stack([np.ones_like(r_prev), r_prev])
        beta_hat = np.linalg.lstsq(X, dr, rcond=None)[0]

        alpha = beta_hat[0]
        beta = beta_hat[1]

        # Convert to Vasicek params
        kappa = -beta / dt  # mean reversion speed
        kappa = max(kappa, 0.01)  # ensure positive

        theta = -alpha / (beta + 1e-12)  # long-run rate
        theta = np.clip(theta, 0.01, 0.20)

        residuals = dr - X @ beta_hat
        if model_type == "cir":
            # CIR: vol proportional to sqrt(rate)
            sigma = float(np.std(residuals / np.sqrt(np.abs(r_prev) + 1e-10)) / np.sqrt(dt))
        else:
            sigma = float(np.std(residuals) / np.sqrt(dt))

        sigma = max(sigma, 0.001)

        # Generate forecast curve
        curve = []
        r_current = rates[-1]
        r = r_current

        for m in range(1, horizon_months + 1):
            # Expected rate under Vasicek/CIR
            r_expected = theta + (r - theta) * np.exp(-kappa * m * dt)
            curve.append({
                "month": m,
                "rate": round(float(r_expected * 100), 3),  # back to %
            })

        # Current rate
        current_rate = round(float(rates[-1] * 100), 3)
        long_run_rate = round(float(theta * 100), 3)

        return {
            "current_rate": current_rate,
            "long_run_rate": long_run_rate,
            "forecast_curve": curve,
            "kappa": round(float(kappa), 4),
            "theta": long_run_rate,
            "sigma": round(float(sigma * 100), 4),  # in %
            "model": model_type.upper(),
            "n_obs": len(rate_history),
            "rate_direction": "rising" if curve[-1]["rate"] > current_rate else "falling",
            "error": None,
        }
    except Exception as e:
        logger.error(f"Vasicek error: {e}")
        return _empty_result(str(e))


def get_risk_free_rate(rate_history: list[float]) -> float:
    """Extract current risk-free rate from RBI repo history for BSM input."""
    if not rate_history:
        return 0.065  # default RBI repo
    return rate_history[-1] / 100


def _empty_result(error: str) -> dict:
    return {
        "current_rate": 0, "long_run_rate": 0,
        "forecast_curve": [],
        "kappa": 0, "theta": 0, "sigma": 0,
        "model": "VASICEK", "n_obs": 0,
        "rate_direction": "stable",
        "error": error,
    }

File: quantfusion/src/models/vasicek.ts
// Vasicek / CIR short-rate model — mean-reverting interest rate

export interface VasicekOutput {
    forecastCurve: { month: number; rate: number }[];
    meanRevSpeed: number;
    longRunRate: number;
}

export function vasicek(
    rateHistory: { date: string; rate: number }[],
    forecastMonths = 12,
): VasicekOutput {
    if (rateHistory.length < 5) {
        return { forecastCurve: [], meanRevSpeed: 0, longRunRate: 0 };
    }

    const rates = rateHistory.map(r => r.rate / 100);
    const n = rates.length;

    // Estimate mean-reversion parameters via OLS on r(t+1) - r(t) = a(b - r(t))dt + noise
    // dr = kappa * (theta - r) * dt
    let sumR = 0, sumR2 = 0, sumDr = 0, sumRDr = 0;
    for (let i = 0; i < n - 1; i++) {
        const dr = rates[i + 1] - rates[i];
        sumR += rates[i];
        sumR2 += rates[i] ** 2;
        sumDr += dr;
        sumRDr += rates[i] * dr;
    }

    const N = n - 1;
    const dt = 1 / 12; // monthly data
    const meanR = sumR / N;

    // OLS: dr = alpha + beta * r → kappa = -beta/dt, theta = -alpha/(beta)
    const beta = (N * sumRDr - sumR * sumDr) / (N * sumR2 - sumR * sumR);
    const alpha = (sumDr - beta * sumR) / N;

    let kappa = Math.max(-beta / dt, 0.01);
    let theta = Math.max(-alpha / (beta || 1e-6), 0.001);

    // Clamp to reasonable values
    kappa = Math.min(kappa, 5);
    theta = Math.min(theta, 0.15);

    // Estimate sigma from residuals
    let ssResid = 0;
    for (let i = 0; i < N; i++) {
        const predicted = alpha + beta * rates[i];
        ssResid += (rates[i + 1] - rates[i] - predicted) ** 2;
    }
    const sigma = Math.sqrt(ssResid / N / dt);

    // Generate forecast curve via Euler discretisation
    let currentRate = rates[rates.length - 1];
    const forecastCurve: { month: number; rate: number }[] = [
        { month: 0, rate: parseFloat((currentRate * 100).toFixed(3)) },
    ];

    for (let m = 1; m <= forecastMonths; m++) {
        currentRate += kappa * (theta - currentRate) * dt;
        // Add small randomness for realistic-looking curve
        currentRate += sigma * Math.sqrt(dt) * (Math.random() * 0.2 - 0.1);
        currentRate = Math.max(currentRate, 0);
        forecastCurve.push({
            month: m,
            rate: parseFloat((currentRate * 100).toFixed(3)),
        });
    }

    return {
        forecastCurve,
        meanRevSpeed: parseFloat(kappa.toFixed(4)),
        longRunRate: parseFloat((theta * 100).toFixed(3)),
    };
}
